File: Silver III/BOJ_1874/BOJ_1874.cpp
#include <algorithm>
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

int n;
stack<int> st;
vector<int> info, numList;
vector<char> answer;

int main()
{
    cin >> n;
    info.resize(n);
    numList.resize(n);
    for (int i = 0; i < n; ++i) {
        cin >> info[i];
    }
    for (int i = 0; i < n; ++i) {
        numList[i] = i + 1;
    }

    int infoIdx = 0;
    int numListIdx = 1;
    st.push(numList[0]);
    answer.push_back('+');
    while (infoIdx < n) {
        if (st.empty() || st.top() < info[infoIdx]) {
            st.push(numList[numListIdx]);
            ++numListIdx;
            answer.push_back('+');
        }
        else if (st.top() == info[infoIdx]) {
            while (!st.empty() && infoIdx < n && st.top() == info[infoIdx]) {
                st.pop();
                ++infoIdx;
                answer.push_back('-');
            }
        }
        else {
            answer.clear();
            break;
        }
    }

    if (!answer.empty()) {
        int answerSize = answer.size();
        for (int i = 0; i < answerSize; ++i) {
            cout << answer[i] << '\n';
        }
    }
    else {
        cout << "NO" << '\n';
    }

    return 0;
}

File: Silver III/BOJ_1874/BOJ_1874.py
import sys
from collections import deque

if __name__ == "__main__":
    n = int(sys.stdin.readline())
    info = [int(sys.stdin.readline()) for _ in range(n)]

    stack = deque()
    start = [i for i in range(1, n + 1)]

    answer = []

    infoIdx = 0
    startIdx = 0
    while True:
        # print(infoIdx, startIdx)
        if not stack:
            if startIdx == n:
                break
            stack.append(start[startIdx])
            startIdx += 1
            answer.append("+")
        else:
            if infoIdx == n:
                break
            value = stack.pop()
            if value == info[infoIdx]:
                infoIdx += 1
                answer.append("-")
            else:
                stack.append(value)
                if startIdx == n:
                    break
                stack.append(start[startIdx])
                startIdx += 1
                answer.append("+")

    if len(answer) != n * 2:
        print("NO")
    else:
        print(*answer, sep="\n")
